core/storage/suggestion/gae_models.py

```diff
@@ -0,0 +1,267 @@
+]
+
+#The delimiter to be used in the suggestion ID.
+SUGGESTION_ID_DELIMITER = '.'
+
+
+class SuggestionParams(ndb.Model):
+    """Model to store all the suggestion params for a suggestion."""
+
+    # The contribution type.
+    contribution_type = ndb.StringProperty(
+        required=True, choices=CONTRIBUTION_TYPE_CHOICES)
+    # The contribution domain. This parameter along with the above contribution
+    # type will be used to determine the category to 'score' the suggester.
+    contribution_domain = ndb.StringProperty(required=True)
+    # The ID of the entity being suggested to.
+    entity_id = ndb.StringProperty(required=True)
+    # The version number of the entity being suggested to.
+    entity_version = ndb.IntegerProperty(required=True)
+
+
+class SuggestionModel(base_models.BaseModel):
+    """Model to store suggestions made by Oppia users."""
+
+    # The type of suggestion.
+    suggestion_type = ndb.StringProperty(
+        required=True, indexed=True, choices=SUGGESTION_TYPE_CHOICES)
+    # The type of entity which the suggestion is linked to.
+    entity_type = ndb.StringProperty(
+        required=True, indexed=True, choices=ENTITY_TYPE_CHOICES)
+    # Status of the suggestion.
+    status = ndb.StringProperty(
+        required=True, indexed=True, choices=STATUS_CHOICES)
+    # Additional parameters to be stored.
+    suggestion_params = ndb.StructuredProperty(SuggestionParams)
+    # The author of the suggestion.
+    author_id = ndb.StringProperty(required=True, indexed=True)
+    # The reviewer who accepted the suggestion.
+    reviewer_id = ndb.StringProperty(required=False, indexed=True)
+    # The thread linked to this suggestion.
+    thread_id = ndb.StringProperty(required=True, indexed=True)
+    # The reviewer assigned to review the suggestion.
+    assigned_reviewer_id = ndb.StringProperty(required=True, indexed=True)
+    # The suggestion payload. Contains the data of the suggestion. For "edit"
+    # suggestions, this contains the change_list. For "add" suggestions, this
+    # contains the data necessary to add the new entity.
+    payload = ndb.JsonProperty(required=True, indexed=False)
+
+    @classmethod
+    def get_instance_id(
+            cls, entity_type, thread_id, entity_id):
+        """Concatenates various parameters and gives the ID of the suggestion
+        model.
+
+        Args:
+            entity_type: str. The type of entity being edited/added.
+            thread_id: str. The ID of the feedback thread linked to the
+                suggestion.
+            entity_id: str. The ID of the entity being edited.
+
+        Returns:
+            str. The full instance ID for the suggestion.
+        """
+        return SUGGESTION_ID_DELIMITER.join([entity_type, thread_id, entity_id])
+
+    @classmethod
+    def create(
+            cls, suggestion_type, entity_type, status,
+            suggestion_params, author_id, reviewer_id, thread_id,
+            assigned_reviewer_id, payload):
```

Comment:
So what's the workflow here exactly -- is the feedback thread created first, and then the suggestion model created after that? Just making sure you have the thread ID by the time you create this model.

If the above is the case -- is there any dependency of the feedback thread model on the suggestion model (e.g. does it need the suggestion model instance ID)? If so, then how would you address that?

Reply:
Yes, the feedback thread is created first, followed by the suggestion model. There is a has_suggestion parameter in the feedback thread model which we will set to true.